File: sdks/java/src/main/java/com/digows/whatsappgateway/command/OutboundCommandResult.java
package com.digows.whatsappgateway.command;

import com.digows.whatsappgateway.operational.SessionReference;

import java.util.Map;

/**
 * Generic execution result published by the gateway for outbound command families.
 * The optional {@code data} map carries family/action-specific return values.
 */
public record OutboundCommandResult(
  String commandId,
  SessionReference session,
  String family,
  String action,
  OutboundCommandResultStatus status,
  String timestamp,
  String reason,
  Map<String, Object> data
)
{
}

File: runtime/src/domain/entities/command/OutboundCommandResult.ts
import { SessionReference } from '../operational/SessionReference.js';
import { OutboundCommandFamily } from './OutboundCommand.js';

export enum OutboundCommandResultStatus {
  Succeeded = 'succeeded',
  Failed = 'failed',
  Blocked = 'blocked',
}

export function parseOutboundCommandResultStatus(value: string): OutboundCommandResultStatus {
  switch (value) {
    case OutboundCommandResultStatus.Succeeded:
      return OutboundCommandResultStatus.Succeeded;
    case OutboundCommandResultStatus.Failed:
      return OutboundCommandResultStatus.Failed;
    case OutboundCommandResultStatus.Blocked:
      return OutboundCommandResultStatus.Blocked;
    default:
      throw new Error(`Unsupported outbound command result status "${value}".`);
  }
}

/**
 * Generic execution result for outbound commands that are not plain message deliveries.
 * Result-specific data is intentionally explicit but extensible through the `data` record.
 */
export class OutboundCommandResult {
  constructor(
    public readonly commandId: string,
    public readonly session: SessionReference,
    public readonly family: OutboundCommandFamily,
    public readonly action: string,
    public readonly status: OutboundCommandResultStatus,
    public readonly timestamp: string,
    public readonly reason?: string,
    public readonly data?: Readonly<Record<string, unknown>>,
  ) {}
}
